File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/PmdDataFrame.swift

import Foundation

struct PmdDataFrame {
    let measurementType: PmdMeasurementType
    let timeStamp: UInt64
    let frameType: PmdDataFrameType
    let isCompressedFrame: Bool
    let dataContent: Data
    let previousTimeStamp: UInt64
    let factor: Float
    let sampleRate: UInt
    
    private static let DELTA_FRAME_BIT_MASK: UInt8 = 0x80
    
    init(data: Data,
         _ getPreviousTimeStamp: (PmdMeasurementType) -> UInt64,
         _ getFactor: (PmdMeasurementType) -> Float,
         _ getSampleRate: (PmdMeasurementType) -> UInt) throws {
        
        measurementType = PmdMeasurementType.fromId(id: data[0])
        let timeBytes = data.subdata(in: 1..<9) as NSData
        var tempTimeStamp: UInt64 = 0
        memcpy(&tempTimeStamp, timeBytes.bytes, 8)
        timeStamp = tempTimeStamp
        
        let frameTypeByte = data[9]
        frameType = try PmdDataFrameType.getTypeFromDataFrameByte(byte: frameTypeByte)
        isCompressedFrame = PmdDataFrame.isCompressedFrame(byte: frameTypeByte)
        dataContent = data.subdata(in: 10..<data.count)
        
        previousTimeStamp = getPreviousTimeStamp(measurementType)
        factor = getFactor(measurementType)
        sampleRate = getSampleRate(measurementType)
    }
    
    private static func isCompressedFrame(byte: UInt8) -> Bool {
        return (byte & PmdDataFrame.DELTA_FRAME_BIT_MASK) > 0
    }

}

enum PmdDataFrameType: UInt8, CaseIterable {
    case type_0 = 0
    case type_1 = 1
    case type_2 = 2
    case type_3 = 3
    case type_4 = 4
    case type_5 = 5
    case type_6 = 6
    case type_7 = 7
    case type_8 = 8
    case type_9 = 9
    
    private static let DATA_FRAME_BIT_MASK: UInt8 = 0x7F
    
    static func getTypeFromDataFrameByte(byte: UInt8) throws -> PmdDataFrameType {
        for type in PmdDataFrameType.allCases {
            if (type.rawValue == (byte & PmdDataFrameType.DATA_FRAME_BIT_MASK)) {
                return type
            }
        }
        throw BleGattException.gattDataError(description: "FrameType id:\(byte) is not implemented")
    }
}

File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/PmdDataFrame.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd

import com.polar.androidcommunications.api.ble.exceptions.BleNotImplemented
import com.polar.androidcommunications.common.ble.TypeUtils

class PmdDataFrame(
    data: ByteArray,
    getPreviousTimeStamp: (PmdMeasurementType, PmdDataFrameType) -> ULong,
    getFactor: (PmdMeasurementType) -> Float,
    getSampleRate: (PmdMeasurementType) -> Int,
) {
    companion object {
        private const val DELTA_FRAME_BIT_MASK = 0x80.toByte()
        private const val FRAME_TYPE_BIT_MASK = 0x7F.toByte()
    }

    val measurementType: PmdMeasurementType = PmdMeasurementType.fromId(data[0])
    val timeStamp: ULong = TypeUtils.convertArrayToUnsignedLong(data, 1, 8)
    val frameType: PmdDataFrameType
    val isCompressedFrame: Boolean
    val dataContent: ByteArray

    val previousTimeStamp: ULong
    val factor: Float
    val sampleRate: Int

    init {
        val frameTypeField = data[9].toUByte()
        frameType = PmdDataFrameType.getTypeFromFrameDataByte(frameTypeField)
        isCompressedFrame = isCompressedFrame(frameTypeField)
        val content = ByteArray(data.size - 10)
        System.arraycopy(data, 10, content, 0, content.size)
        dataContent = content

        previousTimeStamp = getPreviousTimeStamp(measurementType, frameType)
        factor = getFactor(measurementType)
        sampleRate = getSampleRate(measurementType)
    }

    private fun isCompressedFrame(frameTypeByte: UByte): Boolean {
        return frameTypeByte and DELTA_FRAME_BIT_MASK.toUByte() > 0u
    }

    enum class PmdDataFrameType(val id: UByte) {
        TYPE_0(0u),
        TYPE_1(1u),
        TYPE_2(2u),
        TYPE_3(3u),
        TYPE_4(4u),
        TYPE_5(5u),
        TYPE_6(6u),
        TYPE_7(7u),
        TYPE_8(8u),
        TYPE_9(9u),
        TYPE_10(10u);

        companion object {
            fun getTypeFromFrameDataByte(byte: UByte): PmdDataFrameType {
                for (type in values()) {
                    if (type.id == (byte and FRAME_TYPE_BIT_MASK.toUByte())) {
                        return type
                    }
                }
                throw BleNotImplemented("PmdFrameType cannot be parsed from $byte")
            }
        }
    }
}